pkg/statedb2/db.go

```diff
@@ -0,0 +1,233 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package statedb2
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"runtime"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	iradix "github.com/hashicorp/go-immutable-radix/v2"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// DB provides an in-memory transaction database built on top of immutable radix
+// trees. The database supports multiple tables, each with one or more user-defined
+// indexes. Readers can access the data locklessly with a simple atomic pointer read
+// to obtain a snapshot. On writes to the database table-level locks are acquired
+// on target tables and on write transaction commit a root lock is taken to swap
+// in the new root with the modified tables.
+//
+// As data is stored in immutable data structures any objects inserted into
+// it MUST NOT be mutated afterwards.
+//
+// DB holds the "root" tree of tables with each table holding a tree of indexes:
+//
+//	           root
+//	          /    \
+//	         ba    T(foo)
+//	       /   \
+//	   T(bar)  T(baz)
+//
+//	      T(bar).indexes
+//		   /  \
+//		  i    I(byRevision)
```

Comment:
I get why we store tables in the trie, why do we need to store indices in one as well?

Reply:
We store the actual data in immutable radix tries which allow us to do read snapshots. But since we have multiple indecies and multiple tables which should all be part of the same read snapshot, we place the smaller tries in larger tries, until we have 1 single root node for the whole StateDB so we can get a transaction with a single atomic pointer read and update with a single atomic pointer write.

Reply:
I also said in an earlier comment that this can be optimized a lot by just using arrays for tables and indexes. I will follow up with that in a later PR.